py/server/deephaven/table.py

```diff
@@ -407,19 +408,78 @@ def _sort_column(col, dir_):
         _JColumnName.of(col)))
 
 
-def _td_to_columns(table_definition):
-    cols = []
-    j_cols = table_definition.getColumnsArray()
-    for j_col in j_cols:
-        cols.append(
-            Column(
-                name=j_col.getName(),
-                data_type=dtypes.from_jtype(j_col.getDataType()),
-                component_type=dtypes.from_jtype(j_col.getComponentType()),
-                column_type=ColumnType(j_col.getColumnType()),
-            )
+class TableDefinition(JObjectWrapper, Mapping):
+    """A Deephaven table definition."""
+
+    j_object_type = _JTableDefinition
+
+    @staticmethod
+    def from_columns(
+            columns: Union[Mapping[str, DType], Iterable[Column]],
+    ) -> TableDefinition:
+        columns = (
+            [
+                Column(name=name, data_type=data_type)
+                for (name, data_type) in columns.items()
+            ]
+            if isinstance(columns, Mapping)
+            else columns
+        )
+        return TableDefinition(
+            _JTableDefinition.of([col.j_column_definition for col in columns])
         )
-    return cols
+
+    def __init__(self, j_table_definition: jpy.JType):
+        self.j_table_definition = j_table_definition
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self.j_table_definition
+
+    @property
+    def table(self) -> Table:
+        """This table definition as a table."""
+        return Table(_JTableTools.metaTable(self.j_table_definition))
+
+    @cached_property
+    def _dict(self) -> Dict[str, Column]:
+        return {
+            col.name: col
+            for col in [
+                Column(j_column_definition=j_col)
+                for j_col in self.j_table_definition.getColumnsArray()
```

Comment:
If it's not inconvenient to use a `List` (`getColumns()`), doing so would save allocation on the Java side, but I guess this is likely better JNI-wise.

Reply:
Opting not to make any changes here at this time.